### ISSUE TITLE
[BUG] How to set sqlite database backend mlflow in computing cluster?

### ISSUE THREAD

**Issue by Jingnan-Jia** [mlflow/mlflow#5917]
### Discussed in https://github.com/mlflow/mlflow/discussions/5908

<div type='discussions-op-text'>

<sup>Originally posted by **Jingnan-Jia** May 19, 2022</sup>
I use cluster (10 GPU computing nodes, 20 CPU computing nodes, 1 login nodes, 1 management node, with the shared file system) in my daily life. I found mlflow is very useful and convenient to track the experiments run on different computing nodes. But when I have more and more experiments/runs, I found that the mlflow web UI  becomes very slow. This is because the default storage method is file storage system. So the mlflow has to iterately search every directory and read the files. I think database based storage method can speed up it. So I want to set sqlite database as the backend of mlflow. Then I met some difficulties. 

If I want to use the default file storage system, I 
1. do not need to set `mlflow.set_tracking_uri()` in my python script,
2. do not need to run `mlflow server <args>` in my linux terminal,
3. use `mlflow.log_metrics()` and other log functions in my python script and then run `mlflow ui` in my linux terminal. 
Then I can successfully view the experiments records on `localhost: 5000` (I set the port local forward to view the ui in my local PC).

But if I want to use sqlite database backend, these are the steps I applied:
1. set `mlflow.set_tracking_uri()` in my python script,
2. run `mlflow server <args>` in my linux terminal,
3. use `mlflow.log_metrics()` and other log functions in my python script. 
4. Do not need to run `mlflow ui` in my linux terminal, because `mlflow server <>` has coved the function of `mlflow ui` I think.

The key issue here is how to set  `mlflow.set_tracking_uri()` in my python script, and how to set `mlflow server <args>` in my linux terminal. 

At first, I run the following line in my linux terminal on my **login node**:
`mlflow server --backend-store-uri=sqlite:///mlrunsdb.db --default-artifact-root=file:mlruns`. I hope the `Parameters`, `Metrics`, and other small files are saved in `mlrunsdb.db`, and big files like `artifacts` and `model` are saved in file directory `mlruns`.**Is my command right?**

Then, I need to set mlflow.set_tracking_uri. From the [docs](https://www.mlflow.org/docs/latest/python_api/mlflow.html#:~:text=mlflow.set_tracking_uri(uri,%5Bsource%5D), there are 4 options:
1. An empty string, or a local file path, prefixed with file:/. Data is stored locally at the provided file (or ./mlruns if empty).
2. An HTTP URI like https://my-tracking-server:5000.
3. A Databricks workspace, provided as the string “databricks” or, to use a Databricks CLI [profile](https://github.com/databricks/databricks-cli#installation), “databricks://<profileName>”.
4. A pathlib.Path instance

Now the question is how to set it in a sqlite backend mlflow in distriuted cluster? I tried several settings:
At first, do not set `mlflow.set_tracking_uri()` at all just as what I have done for the file storage system before. The result is that the code can be run successfully but the web ui did not record anything. In contrast, when I terminated the command `mlflow server --backend-store-uri=sqlite:///mlrunsdb.db --default-artifact-root=file:mlruns` using `ctrl+c`, then run `mlflow ui`, the recorded runs/experiments could be viewed now! So this means that I have to set `mlflow.set_tracking_uri()` if I want the mlflow to record parameters to database instead of default file system.

- So then I tried to set uri as `mlflow.set_tracking_uri("http://localhost:5000")`.
Then I noticed that if I ran the code on different computing nodes, these experiments all failed because of the error: `ConnectionRefusedError: [Errno 111] Connection refused`. But if I ran the code on the login node, it was okay. I guess this is because the `localhost` in my `set_tracking_uri("http://localhost:5000")` means the localhost of the login node instead of any other computing nodes. **Is it right?**

- So I tried to set `mlflow.set_tracking_uri("http://10.161.27.235:5000")` (Here 10.161.27.235 is the IP address of the login node). But the code can not be run successfully at all no matter on computing nodes or login node. The error is: `requests.exceptions.ConnectionError: HTTPConnectionPool(host='10.161.27.235', port=5000): Max retries exceeded with url: /api/2.0/mlflow/experiments/list?view_type=ALL (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x2aabd74e12e0>: Failed to establish a new connection: [Errno 111] Connection refused'))`
So my conclusion is that maybe I should not set HTTP URI for such a computing cluster. **Is my conclusion right?**

- So I tried `mlflow.set_tracking_uri('sqlite:///mlrunsdb2.db')` (before that I ran `mlflow server --backend-store-uri=sqlite:///mlrunsdb2.db --default-artifact-root=file:mlruns` on my linux terminal in login node). 
The result was very weird. 
  1. The code can be run successfully on computing node for several minute, but when I refresh the web ui or just no reason, the experiment would broke down immediatly. The error information is: `mlflow.exceptions.MlflowException: (sqlite3.OperationalError) disk I/O error (Background on this error at: https://sqlalche.me/e/14/e3q8) `
  2. The code can be run successfully on login node, even refreshing the web ui can not lead it to break down. But when I launched another experiment in a computing node, the two experiments (one in login node, anoter one in computing node) will break down at the same time because of unclear reason (error is: `mlflow.exceptions.MlflowException: (sqlite3.OperationalError) disk I/O error
(Background on this error at: https://sqlalche.me/e/14/e3q8)`).

To summarize, I tried `set_tracking_uri` using HTTP URL or `sqlite:///mlrunsdb2.db`, but both failed. Errors are different. 

**Can you tell me what is the correct way to achieve sqlite database backend mlflow in computing cluster?**

</div>

**Comment by BenWilson2** [comment]
Could you try starting the tracking server's host as, instead of (default) `--host localhost --port 5000` instead set it to `--host 0.0.0.0 --port 5000` so that multiple machines can interface with the server on that port?

**Comment by Jingnan-Jia** [comment]
@BenWilson2 It does not work.

If I set `mlflow server --backend-store-uri=sqlite:///mlrunsdb15.db --default-artifact-root=file:mlruns --host 0.0.0.0 --port 5000` on the linux terminal, I have two choices in python script:
1. `mlflow.set_tracking_uri('http://localhost:5000')`, then error was 
```Traceback (most recent call last):
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/urllib3/connection.py", line 174, in _new_conn
    conn = connection.create_connection(
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/urllib3/util/connection.py", line 95, in create_connection
    raise err
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/urllib3/util/connection.py", line 85, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused
...........
.........
........
Traceback (most recent call last):
  File "run.py", line 165, in <module>
    mlflow.set_experiment("lung_fun_db15")
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/mlflow/tracking/fluent.py", line 110, in set_experiment
    experiment = client.get_experiment_by_name(experiment_name)
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/mlflow/tracking/client.py", line 462, in get_experiment_by_name
    return self._tracking_client.get_experiment_by_name(name)
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/mlflow/tracking/_tracking_service/client.py", line 162, in get_experiment_by_name
    return self.store.get_experiment_by_name(name)
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/mlflow/store/tracking/rest_store.py", line 297, in get_experiment_by_name
    for experiment in self.list_experiments(ViewType.ALL):
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/mlflow/store/tracking/rest_store.py", line 77, in list_experiments
    response_proto = self._call_endpoint(ListExperiments, req_body)
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/mlflow/store/tracking/rest_store.py", line 56, in _call_endpoint
    return call_endpoint(self.get_host_creds(), endpoint, method, json_body, response_proto)
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/mlflow/utils/rest_utils.py", line 249, in call_endpoint
    response = http_request(
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/mlflow/utils/rest_utils.py", line 163, in http_request
    raise MlflowException("API request to %s failed with exception %s" % (url, e))
mlflow.exceptions.MlflowException: API request to http://localhost:5000/api/2.0/mlflow/experiments/list failed with exception HTTPConnectionPool(host='localhost', port=5000): Max retries exceeded with url: /api/2.0/mlflow/experiments/list?view_type=ALL (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x2aabd74d7400>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

2. `mlflow.set_tracking_uri('sqlite:///mlrunsdb15.db')`, then the code can run for several minutes but at some time the error occured: 
```
Traceback (most recent call last):
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1015, in _commit_impl
    self.engine.dialect.do_commit(self.connection)
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 685, in do_commit
    dbapi_connection.commit()
sqlite3.OperationalError: disk I/O error

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/mlflow/store/db/utils.py", line 79, in make_managed_session
    yield session
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/mlflow/store/tracking/sqlalchemy_store.py", line 633, in _log_metrics
    _insert_metrics(metric_instances)
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/mlflow/store/tracking/sqlalchemy_store.py", line 630, in _insert_metrics
    session.commit()
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 1431, in commit
    self._transaction.commit(_to_root=self.future)
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 836, in commit
    trans.commit()
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 2378, in commit
    self._do_commit()
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 2568, in _do_commit
    self._connection_commit_impl()
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 2539, in _connection_commit_impl
    self.connection._commit_impl()
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1017, in _commit_impl
    self._handle_dbapi_exception(e, None, None, None, None)
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 2043, in _handle_dbapi_exception
    util.raise_(
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1015, in _commit_impl
    self.engine.dialect.do_commit(self.connection)
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 685, in do_commit
    dbapi_connection.commit()
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) disk I/O error
(Background on this error at: https://sqlalche.me/e/14/e3q8)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "run.py", line 174, in <module>
    run(args)
  File "run.py", line 144, in run
    best_valid_loss = step('valid', net, data_dt['valid'], loss_fun, opt, i, target, mypath, False, best_valid_loss)
  File "run.py", line 84, in step
    log_metric(mode+'MAEBatch_All', mae_all, data_idx+epoch_idx*len(dataloader))
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/mlflow/tracking/fluent.py", line 583, in log_metric
    MlflowClient().log_metric(run_id, key, value, int(time.time() * 1000), step or 0)
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/mlflow/tracking/client.py", line 690, in log_metric
    self._tracking_client.log_metric(run_id, key, value, timestamp, step)
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/mlflow/tracking/_tracking_service/client.py", line 226, in log_metric
    self.store.log_metric(run_id, metric)
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/mlflow/store/tracking/sqlalchemy_store.py", line 598, in log_metric
    self._log_metrics(run_id, [metric])
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/mlflow/store/tracking/sqlalchemy_store.py", line 667, in _log_metrics
    _insert_metrics(non_existing_metrics)
  File "/home/jjia/.conda/envs/py38/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/jjia/.conda/envs/py38/lib/python3.8/site-packages/mlflow/store/db/utils.py", line 86, in make_managed_session
    raise MlflowException(message=e, error_code=BAD_REQUEST)
mlflow.exceptions.MlflowException: (sqlite3.OperationalError) disk I/O error
(Background on this error at: https://sqlalche.me/e/14/e3q8)

```

**Comment by Jingnan-Jia** [comment]
@BenWilson2 Thanks ! It started to work when I set `mlflow.set_tracking_uri('http://login-ip-address:5000')`

**Comment by BenWilson2** [comment]
Awesome! Glad that you're up and running! :)
